### PR TITLE
V2.0

### PR DESCRIPTION
**Pull request by teyeheimans**
- Updated minimal supported PHP version to 7.4, which allows us to use property type hinting, short function syntax and more.
- Removed DRUID_VERSION config setting. This was only used for [whereFlags()](README.md#whereflags) and made our code ugly. 
  You can now use the oldskool javascript variant by setting the new 4th parameter to true.
- Added `markAsUnused` option to kill task and the KillTaskBuilder.
- Refactored the IndexTask. It now allows an InputSource. Also added option to specify timestamp column. Added support
  to be able to ingest spatial dimensions using the [spatialDimension()](README.md#importing-data-using-a-batch-index-job) 
  method in the IndexTaskBuilder.
- Added spatial filter methods:
    - [whereSpatialRectangular()](README.md#wherespatialrectangular)
    - [whereSpatialRadius()](README.md#wherespatialradius)
    - [whereSpatialPolygon()](README.md#wherespatialpolygon)
    - [orWhereSpatialRectangular()](README.md#orwherespatialrectangular)
    - [orWhereSpatialRadius()](README.md#orwherespatialradius)
    - [orWhereSpatialPolygon()](README.md#orwherespatialpolygon)
- Added ExpressionFilter and the [whereExpression()](README.md#whereexpression) and [orWhereExpression()](README.md#orwhereexpression) methods.
- Removed IngestSegmentFirehose and FirehoseInterface. These are now replaced by InputSources.
- Added _a lot_ of [input sources](README.md#input-sources) (used for index tasks):
    - Azure
    - Google Cloud
    - S3
    - HDFS
    - Http
    - Local
    - Inline
    - SQL
    - Combine
- Added [input formats](README.md#input-formats) (used for index tasks):
    - csv Format
    - tsv Format
    - json Format
    - orc Format
    - parquet Format
    - protobuf Format
- Updated DruidInputSource so that you now can also specify a filter on it.
- Added [pollTaskStatus()](README.md#druidclientpolltaskstatus) in the client, which will poll until the status of a task is other than `RUNNING`.
  The time between each check can be influenced by the `'polling_sleep_seconds'` config setting or
  the `DRUID_POLLING_SLEEP_SECONDS` .env setting for Laravel/Lumen applications.
- Added support to join other DataSources. See: [Data Sources](README.md#querybuilder-data-sources)
- Added support for `inlineLookup`. See [inlineLookup()](README.md#inlinelookup)
- Added support for multi-value columns. Either by using one of the new [multiValueListSelect](README.md#multivaluelistselect), 
  [multiValueRegexSelect](README.md#multivalueregexselect), and [multiValuePrefixSelect](README.md#multivalueprefixselect) dimension selector filters. 
- The `$dataSource` parameter for the [query()](README.md#druidclientquery) method on the druid-client is now optional. You can specify later on the
  query builder which dataSource you want to use.

### CODE REVIEW

File: src/Tasks/CompactTask.php [hunk 1]
```diff
@@ -71,7 +51,7 @@ class CompactTask implements TaskInterface
     public function __construct(
         string $dataSource,
         IntervalInterface $interval,
-        $segmentGranularity = null,
+        string $segmentGranularity = null,
```
Comment (reinder83) [inline src/Tasks/CompactTask.php]: This should be `?string $segmentGranularity = null`

**Review by reinder83 — APPROVED** [review]
LGTM!

File: infection.json.dist [hunk 2]
```diff
@@ -5,7 +5,8 @@
             "src"
         ],
         "excludes": [
-            "*DruidService*"
+            "*DruidService*",
+
```
Comment (jasper-ter-veen) [inline infection.json.dist]: Unnecessary white line

File: src/Concerns/HasIntervals.php [hunk 3]
```diff
@@ -49,7 +49,7 @@ trait HasIntervals
      * @return $this
      * @throws \Exception
      */
-    public function interval($start, $stop = null)
+    public function interval($start, $stop = null): self
```
Comment (jasper-ter-veen) [inline src/Concerns/HasIntervals.php:52]: Waarom niet `addInterval`?

File: src/DruidClient.php [hunk 4]
```diff
@@ -331,22 +313,26 @@ protected function makeGuzzleClient(): GuzzleClient
...
+     * @return array<string,array<mixed>|string|int>
      * @throws \Level23\Druid\Exceptions\QueryResponseException
      */
     protected function parseResponse(ResponseInterface $response, array $query = []): array
     {
         $contents = $response->getBody()->getContents();
         try {
-            $row = \json_decode($contents, true) ?: [];
-            if (\JSON_ERROR_NONE !== \json_last_error()) {
+            $row = json_decode($contents, true) ?: [];
```
Comment (jasper-ter-veen) [inline src/DruidClient.php]: Waarom niet JSON_THROW_ON_ERROR ?

File: src/Concerns/HasIntervals.php (hunk 3 above)
Comment (teyeheimans) [reply src/Concerns/HasIntervals.php:52]: Dat zou sowieso een breaking change zijn in *elke* query welke er bestaat. Verder wordt er nergens een "add" prefix gebruikt in de methodes welke gebruikt worden in de "public methods". Ik zie ook niet in waarom dat logischer zou zijn.

File: infection.json.dist (hunk 2 above)
Comment (teyeheimans) [reply infection.json.dist]: Fixed!

File: src/DruidClient.php (hunk 4 above)
Comment (teyeheimans) [reply src/DruidClient.php]: Ja, dat zou kunnen. Alleen ik zie niet perse wat de meerwaarde is van als we dat gaan gebruiken?

File: src/Tasks/TaskBuilder.php [hunk 5]
```diff
@@ -90,10 +94,10 @@ public function toJson($context = []): string
     {
         $task = $this->buildTask($context);
 
-        $json = \json_encode($task->toArray(), JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
-        if (\JSON_ERROR_NONE !== \json_last_error()) {
+        $json = json_encode($task->toArray(), JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
```
Comment (jasper-ter-veen) [inline src/Tasks/TaskBuilder.php]: Gebruik JSON_THROW_ON_ERROR

**Pull request merged**